h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1758,90 +1769,117 @@ private void compactMoveChunks(Iterable<Chunk> move) {
             // to the end of the file
             writeStoreHeader();
             sync();
+
+            long reservedAreaSize = getFileLengthInUse();
             for (Iterator<Chunk> iter = move.iterator(); iter.hasNext(); ) {
                 Chunk chunk = iter.next();
-                if (fileStore.predictAllocation(chunk.len) >= leftmostBlock) {
+                // while chunks destination do not overlap with their
+                // source location as a group, we can move them directly
+                // to the target destination
+                if (fileStore.predictAllocation(chunk.len) + chunk.len > leftmostBlock) {
+                    // overlap encountered, now rest of the chunks
+                    // are going to be moved in two steps - first to EOF,
+                    // and only then to target destination
+                    // this is to avoid possible corruption on abrupt shutdown
                     break;
                 }
-                moveChunk(chunk, false);
+                moveChunk(chunk, 0);
                 iter.remove();
             }
 
+            boolean movedToEOF = false;
+            // we need to ensure that chunks moved within the following loop
+            // do not overlap with space just released by chunks moved above,
+            // if some of them happened to sit right at the end of used space,
+            // hence the need to reserve this area
             for (Chunk chunk : move) {
-                moveChunk(chunk, true);
+                moveChunk(chunk, reservedAreaSize);
+                movedToEOF = true;
             }
 
-            // update the metadata (store at the end of the file)
-            reuseSpace = false;
-            commit();
-            sync();
+            long reservedAreaHigh = getFileLengthInUse();
+            Chunk chunk = null;
+            if (movedToEOF) {
+                // update the metadata (store at the end of the file)
+                reuseSpace = false;
+                commit();
+                sync();
 
-            Chunk chunk = lastChunk;
+                chunk = lastChunk;
+                reservedAreaHigh = getFileLengthInUse();
 
-            // now re-use the empty space
-            reuseSpace = true;
-            for (Chunk c : move) {
-                moveChunk(c, false);
+                // now re-use the empty space
+                reuseSpace = true;
+                for (Chunk c : move) {
+                    moveChunk(c, 0);
+                }
             }
 
             // update the metadata (within the file)
+            reservedLow = movedToEOF ? reservedAreaSize : leftmostBlock * BLOCK_SIZE;
+            reservedHigh = reservedAreaHigh;
             commit();
             sync();
-            if (moveChunk(chunk, false)) {
+            if (chunk != null &&
+                    fileStore.predictAllocation(chunk.len) + chunk.len < chunk.block &&
```

Comment:
In line 1824, it's unnecessary that plus chunk.len, because of the chunk is not moved yet here, and not overlapped. Also we need reserve the area of the two last chunks if they moved.

Reply:
this has been addressed with newer commits